### PR TITLE
Add a number of checks to the builder

### PR DESCRIPTION
**Pull request by kevmoo**
- Validate that adding a `target` to @CloudFunction parses
- Make sure only one function has a given `target`
- Test that all valid function shapes are supported
- Validate that annotated members are functions with the right parameters and return type.

### CODE REVIEW

**Comment by kevmoo** [comment]
@subfuzion – each commit stands by itself pretty nicely. 

Please do a rebase merge – or just let me handle it!

File: functions_framework_builder/lib/builder.dart [hunk 1]
```diff
@@ -65,9 +73,6 @@ class _FunctionsFrameworkBuilder implements Builder {
     );
     await buildStep.writeAsString(mainDart, '''
 // GENERATED CODE - DO NOT MODIFY BY HAND
-// Copyright (c) 2020, the Dart project authors.
```
Comment (subfuzion) [inline functions_framework_builder/lib/builder.dart:68]: Please don't remove this. I had checked with @grant on this [the other day](https://github.com/GoogleCloudPlatform/functions-framework-dart/pull/4#issuecomment-731823640) about ensuring copyrights even in generated code.

File: functions_framework_builder/test/builder_test.dart [hunk 2]
```diff
@@ -14,33 +17,163 @@ void main() {
...
+      'Responose handleGet() => null;': notCompaitbleMatcher,
+      // First param is not positional
+      'Responose handleGet({Request reques}) => null;': notCompaitbleMatcher,
+      // Too many required params
+      'Responose handleGet(Request request, int other) => null;':
+          notCompaitbleMatcher,
+      // Param is wrong type
+      'Responose handleGet(int request) => null;': notCompaitbleMatcher,
+      // Return type is wrong
+      'int handleGet(Request request) => null;': notCompaitbleMatcher,
```
Comment (subfuzion) [inline functions_framework_builder/test/builder_test.dart:119]: s/notCompaitbleMatcher/notCompatibleMatcher/g

File: functions_framework_builder/test/builder_test.dart [hunk 3]
```diff
@@ -62,39 +195,19 @@ const _ignoredLogMessages = {
...
-import 'package:shelf/shelf.dart';
-
-@CloudFunction()
-Response handleGet(Request request) => Response.ok('Hello, World!');
-''';
-
-const _testGenPartContent = r'''
+String get _outputHeader => '''
 // GENERATED CODE - DO NOT MODIFY BY HAND
-// Copyright (c) 2020, the Dart project authors.
```
Comment (subfuzion) [inline functions_framework_builder/test/builder_test.dart:84]: See previous comment about retaining copyright even for generated code.

**Review by subfuzion — COMMENTED** [review]
Looks great! Just one typo and a request to leave the copyright in, even for generated code. I actually don't have an opinion on that, so you may want to discuss offline with @grant. Or if you're certain it's not needed (e.g., no other generated code in the Dart project does this), then make the call and merge.

File: functions_framework_builder/test/builder_test.dart (hunk 2 above)
Comment (subfuzion) [reply functions_framework_builder/test/builder_test.dart:119]: Also just noticed line 117: s/Responose/Response

File: examples/hello/pubspec.yaml [hunk 4]
```diff
@@ -15,3 +15,7 @@ dev_dependencies:
   http: ^0.12.0
   test: ^1.15.0
   test_process: ^1.0.0
+
+dependency_overrides:
```
Comment (subfuzion) [inline examples/hello/pubspec.yaml:19]: Even though it's part of the repo, I'd prefer the example to remain standalone. Let's discuss offline before going further. Thanks!

**Review by subfuzion — DISMISSED** [review]

File: functions_framework_builder/test/builder_test.dart (hunk 3 above)
Comment (kevmoo) [reply functions_framework_builder/test/builder_test.dart:84]: ack

File: functions_framework_builder/lib/builder.dart (hunk 1 above)
Comment (kevmoo) [reply functions_framework_builder/lib/builder.dart:68]: ack

File: functions_framework_builder/test/builder_test.dart (hunk 3 above)
Comment (kevmoo) [reply functions_framework_builder/test/builder_test.dart:84]: ack

File: examples/hello/pubspec.yaml (hunk 4 above)
Comment (kevmoo) [reply examples/hello/pubspec.yaml:19]: ack

File: functions_framework_builder/test/builder_test.dart (hunk 2 above)
Comment (kevmoo) [reply functions_framework_builder/test/builder_test.dart:119]: ack

**Review by subfuzion — APPROVED** [review]

**Pull request merged**